### DISCUSSION TITLE
How to access novias node relationships data in the labels function?

### DISCUSSION THREAD

**Discussion by ra-haik** [neo4j-contrib/neovis.js#345] [Q&A] [answered]
```js
labels:
            {
                ID: {
                    label: "name",
                    identifier: "ID",
                    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
                        static: {
                            color: "#97c2fc",
                            shape: "dot",
                            size: 5,

                        },
                        function:  {
                            size: (node) => {           
                                return // how to access node.reletioships so that I can count the properties.visits.length like below?
                            }
                        },
                    },
                },
            },
            relationships: {
                TRAFFIC: {
                    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
                        static: {
                            width:1,
                        },
                        function: {
                            width: (node) => {
                                return node.properties.visits.length;
                            }
                        },
                    },
                },          
            },
```

**Comment by thebestnom** [comment]
You should do it in the cypher as neo4j doesn't return a way to get to the relationships in the object

I didn't touch cypher in a long time so need to search how I did it back than

**Reply by thebestnom** [reply]
Also, which relationship? If you have multiple which one do you want?

**Reply by thebestnom** [reply]
https://stackoverflow.com/questions/45837277/is-it-possible-to-add-a-temporary-property-to-a-neo4j-node-for-return-only
This to add property to the return, and do that the property is the data you want from the relationship

**Reply by ra-haik** [reply]
I need to loop over the relationships collection and count the items in the relationship's properties.

**Reply by thebestnom** [reply]
(Also for me I need to add a way to do `a.b` in the mapped config 😅)

**Reply by thebestnom** [reply]
> I need to loop over the relationships collection and count the items in the relationship's properties.

You can pretty easily do that in the cypher

**Reply by thebestnom** [reply]
```cypher
MATCH (a:  label)-[r]->(b)
RETURN a{.*, Temp_size: sum(r.visits.length)}, r
```
Or somthing like that

**Reply by ra-haik** [reply]
> You can pretty easily do that in the cypher

Not "easily", I tried for like 8 hours and now I am looking for answers here.
I was under the impression that the connection between node and a relationship, is something that should be trivial, and so could not believe that I need to write extra code to make it work.
The stack post is from 2017, are you sure it is still relevant?

**Reply by thebestnom** [reply]
Sorry for using the word easily, didn't mean to sound too harsh 😅 and no, the connection between node and relationship outside of the actual graph is not 100% trivial while building the graph 😅
But cypher is exactly where things like that are meant to be used

**Reply by thebestnom** [reply]
The node and the relationships object are separate until graph is completely build in neovis (also in neo4j) and only coupled by ids, so only in the cypher or after the graph is built there is easy access to one another

**Reply by thebestnom** [reply]
> ```cypher
> MATCH (a:  label)-[r]->(b)
> RETURN a{.*, Temp_size: sum(r.visits.length)}, r
> ```
> 
> Or somthing like that

Somthing like that didn't work?

**Reply by ra-haik** [reply]
> > ```cypher
> > MATCH (a:  label)-[r]->(b)
> > RETURN a{.*, Temp_size: sum(r.visits.length)}, r
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or somthing like that
> 
> Somthing like that didn't work?

Not even remotely, when I use the a{.*} thing, the graph stops showing it.

**Reply by thebestnom** [reply]
Ill test that tomorrow, it doesn't even work in 5he neo4j desktop?

**Reply by ra-haik** [reply]
> Ill test that tomorrow, it doesn't even work in 5he neo4j desktop?

No, desktop does not work as well...
```cypher
MATCH(User)- [TRAFFIC] -> (ID) WITH User, TRAFFIC, ID, sum(size(TRAFFIC.visits)) as total_visits RETURN User, TRAFFIC, ID{identity: id(ID), label: labels(ID), properties: properties(ID),elementId: id(ID),visits: total_visits } LIMIT 30
```

**Reply by thebestnom** [reply]
Ok, tested quickly on my phone, yes it shouldn't work because thanks to neo4j it doesnt return as node but as a dictionary because of course it does... But! I think you can maybe use advanced cypher to do that

```ts
{
// Config
    [NeoVis.NEOVIS_ADVANCED_CONFIG]: {
        cypher: {
             size: "match (a)-[r]->() where id(a) = $id return sum(r.visits.length)"
        }
     }
}
```

**Reply by thebestnom** [reply]
Again sorry for using the word easily, it always gets back at you 😅

**Reply by ra-haik** [reply]
> ```ts
> size: "match (a)-[r]->() where id(a) = $id return sum(r.visits.length)"
> ```

This is not a solution, not sure if this works because I get a server error "inefficient resources", not sure what this means.
The problem with this idea is that for every node, this funky thing will make a request to the server, whereas the  information that the graph needs is already downloaded and can be linked to create the relationships in code.

**Reply by thebestnom** [reply]
It is actually not have to be downloaded as most of the time neo4j driver will return the records completly seprated per relationship, and it streams the data to the driver (and neovis.js run each stream, not when the data fully downloaded as it's way faster 😅)
The inefficient resources though is weird as I remember I did that it will not open new connection and reuse the same connection for all the queries

**Reply by thebestnom** [reply]
But I think Im going to add anyway that Ill send the original Record also to the advanced function, so you can add whatever you want to the cypher return and use it, Ill try to do it as fast as possible 😅 but it will take me few hours before Ill be able to touch that

**Reply by ra-haik** [reply]
> It is actually not have to be downloaded as most of the time neo4j driver will return the records completly seprated per relationship, and it streams the data to the driver (and neovis.js run each stream, not when the data fully downloaded as it's way faster 😅) The inefficient resources though is weird as I remember I did that it will not open new connection and reuse the same connection for all the queries

My understanding of all this is very poor, so I can't help in any way.
One thought that I had, you say that you **can't** associate the nodes with the relationship **before** they are drawn. Does this mean that you **can** associate them **after** they are drawn? If this is the case, then will it be possible to add an event-handler to the node, so that when the "done drawing" event is raised, the callback will go and get the information stored inside the relationships, and use it to redraw the node to the desired size/shape/color?
Thank you very much for your help!

**Reply by thebestnom** [reply]
After they are drawn you can do it, you have access to neovis.nodes and neovis.edges that are visData object that you can manipulate and that are linked to the graph (so changing them would change the graph) and they have the raw data on them

**Reply by ra-haik** [reply]
> neovis.nodes and neovis.edges

That is awesome! Is there an example for this, how to write the code? I didn't find anything helpful.

**Reply by thebestnom** [reply]
Umm... Not really, but after the complete event you can to use it directly and if you get stuck you can continue ask here 😅 the tests kinda use it to get data, and `neovis.nodes.update{id: id, newdata}` should be pretty much the api to update it

**Reply by ra-haik** [reply]
> complete event

"complete event" 

> Umm... Not really, but after the complete event you can to use it directly and if you get stuck you can continue ask here 😅 the tests kinda use it to get data, and `neovis.nodes.update{id: id, newdata}` should be pretty much the api to update it

For some reason it is not working for me, the **viz.node** does not contain anything when the event is raised 
```js

      const viz = new NeoVis(config);
        viz.render();

        viz.registerOnEvent('completed', () => {
            console.log("Graph visualization completed.");

            // Update node size based on relationship data

            console.log(viz.edges);

            console.log(viz.nodes); 

            for (let nodeId in viz.nodes) {
        // noting here
            }
        });
```

**Reply by thebestnom** [reply]
Thats... Really weird, it makes no sense (I think maybe you can't iterate over it but it is not empty?)
https://visjs.github.io/vis-data/data/dataset.html maybe try `vis.nodes.forEach`?

**Reply by ra-haik** [reply, marked as answer]
Thank you very much for your help, here is the code that is working.

```js
        const viz = new NeoVis(config);
        viz.render();

        viz.registerOnEvent('completed', () => {
            console.log("Graph visualization completed.");

            viz.nodes.forEach((node) =>
            {
                let newSize = 0;

                if (node.group == 'ActiveDirectoryDomain') {
                    viz.nodes.update({ id: node.id, size: 10 });
                }
                else if (node.group == 'User') {
                    viz.nodes.update({ id: node.id, size: 5 });
                }
                else {

                    viz.edges.forEach((edge) => {
                        if (edge.from === node.id || edge.to === node.id) {
                            newSize += JSON.parse(edge.raw.properties.visits ?? '[]').length;
                        }
                    });

                    viz.nodes.update({ id: node.id, size: newSize });
                }
            }); 
        });
```

**Reply by thebestnom** [reply]
That's really weird you need to parse the array, but other than that Im glad managed, also you can use vis.edges.get({to: node.id}) to get all edges instead of iterating over all of it 😅

**Reply by ra-haik** [reply]
> That's really weird you need to parse the array, but other than that Im glad managed, also you can use vis.edges.get({to: node.id}) to get all edges instead of iterating over all of it 😅

Is it possible to use array type in neo4j, I don't see it on the list?


![image](https://github.com/neo4j-contrib/neovis.js/assets/130375207/5e04714e-508a-462e-ae95-628b1deac182)

**Reply by thebestnom** [reply]
Not sure what app are you using, but in the cypher you can do things like
```cypher
MATCH (a)
SET a.a = [1,2,3,4]
```
And it should work

**Reply by ra-haik** [reply]
> Not sure what app are you using, but in the cypher you can do things like
> 
> ```cypher
> MATCH (a)
> SET a.a = [1,2,3,4]
> ```
> 
> And it should work

Thanks, was able to fix it, the issue was with the data loading tool.
